src/current/advisories/a99561.md

```diff
@@ -0,0 +1,51 @@
+---
+title: Technical Advisory 99561
+advisory: A-99561
+summary: When adding a new column using the new declarative schema changer, CockroachDB retrieves a unique secondary index mistakenly, assuming it to be a primary index.
+toc: true
+affected_versions: v22.2.x, v23.1.0-v23.1.5
+advisory_date: 2023-07-07
+docs_area: releases
+---
+
+Publication date: {{ page.advisory_date | date: "%B %e, %Y" }}
+
+## Description
+
+In CockroachDB v20.2 and v21.1 (the first two versions to support [`ALTER PRIMARY KEY`](../v21.1/alter-primary-key.html)), there was a bug where in some cases the index automatically created to provide uniqueness for the existing primary key would have its `EncodingType` field marked as `PrimaryIndexEncoding`. This behavior was fixed in [#66599](https://github.com/cockroachdb/cockroach/pull/66599), but left secondary indexes with primary encoding.
+
+In CockroachDB v22.2 and later, the implementation for [`ALTER TABLE`](../v22.2/alter-table.html#add-column) began using the new declarative schema changer framework. When adding a new column using the new [declarative schema changer](../v22.2/online-schema-changes.html#declarative-schema-changer), CockroachDB retrieves a unique secondary index mistakenly, assuming it to be a primary index. As a result, the secondary index is modified to store the new columns. CockroachDB expects there to be only one primary index, but in the case outlined here, there are two, because a secondary index has been incorrectly labeled as a primary one. This will cause the new column data to be present only in the secondary index and could lead to irrecoverable data loss, if the secondary index is deleted by the user.
+
+## Statement
+
+This is resolved in CockroachDB by [#105828](https://github.com/cockroachdb/cockroach/pull/105828), which makes a post-deserialization change to alter the encoding version for secondary indexes that are marked as primary index encoded and there are no mutations. This PR will also add validation to detect tables with this same problem so that they can be fixed.
```

Comment:
I get lost in the description of the fix. I think "primary index encoded" might be easier to understand as "encoded as primary indexes", and I'm confused about "and there are no mutations" -- does it mean "where there are no mutations? Secondary indexes that are encoded as primary indexes where there are no mutations? (No mutations on what?)

Reply:
@rimadeodhar Can you clarify the mutations part here? I assumed it meant mutations hadn't occurred on the secondary indexes, but I realize this is unclear.

Reply:
@mdlinville I have updated this language per Rafi for this sentence.